<?xml version="1.0"?>
<launch>
    
    <arg name="mav_name" value="$(optenv MAV_NAME)" />
    <arg name="mav_id" value="$(optenv MAV_ID)" />
    <arg name="mav_cam_name" value="$(optenv MAV_CAM_NAME)" />
    <arg name="log_output_mavros" default="log" />
    <arg name="respawn_mavros" default="false" />
    <arg name="fcu_url_pilot" default="/dev/ttyUSB0:921600" />
    <arg name="gcs_url_pilot" default="udp://@192.168.0.125:14550" />
    <arg name="fcu_protocol" default="v2.0" />
    <arg name="fix_lat" default="26.5093660" />
    <arg name="fix_long" default="80.2269306" />

    <arg name="camera_info_path" value="$(find mav_utils)/cfg/cam/ocam_640x480.yaml" if="$(eval mav_cam_name == 'ocam')" />
    <arg name="camera_info_path" value="$(find mav_utils)/cfg/cam/mv.yaml" if="$(eval mav_cam_name == 'bluefox')" />
    <arg name="image_topic" value="/camera/image_raw" if="$(eval mav_cam_name == 'ocam')" />
    <arg name="image_topic" value="/mv/image_raw" if="$(eval mav_cam_name == 'bluefox')" />

    <arg name="uav2_utm_pose" value="/spedix/utm_pose" if="$(eval mav_name == 'magnus')" />
    <arg name="uav3_utm_pose" value="/vega/utm_pose" if="$(eval mav_name == 'magnus')" />

    <arg name="uav2_utm_pose" value="/magnus/utm_pose" if="$(eval mav_name == 'spedix')" />
    <arg name="uav3_utm_pose" value="/vega/utm_pose" if="$(eval mav_name == 'spedix')" />

    <arg name="uav2_utm_pose" value="/magnus/utm_pose" if="$(eval mav_name == 'vega')" />
    <arg name="uav3_utm_pose" value="/spedix/utm_pose" if="$(eval mav_name == 'vega')" />

    <group ns="$(arg mav_name)">

        <node name="planner" pkg="imav_planner" type="planner" output="screen" clear_params="true">

            <!-- parameters -->
            <rosparam file="$(find imav_planner)/cfg/planner.yaml" command="load"/>

            <!-- subscribers -->
            <remap from="odometry" to="pilot/local_position/odom"/>
            <remap from="drop_info" to="drop_info"/>
            <remap from="home_info" to="home_pose"/>

            <remap from="object_poses" to="object_poses"/>
            <remap from="helipad" to="hdetect/helipad_position"/>
            <remap from="global_pose" to="pilot/global_position/global"/>
            <remap from="mission/reached" to="pilot/mission/reached"/>
            <remap from="state" to="pilot/state" />

            <!-- publishers -->
            <remap from="planner/curr_state" to="planner/curr_state"/>
            <remap from="servo" to="gripper_status"/>
            <remap from="mission_info" to="mission_setpoint"/>

            <!-- service clients -->
            <!-- <remap from="back_to_position_hold" to="back_to_position_hold"/> -->
            <remap from="set_mode" to="pilot/set_mode"/>
            <remap from="mission/pull" to="pilot/mission/pull"/>
            <remap from="detector/terminate" to="detector/terminate"/>
            <remap from="hdetect/terminate" to="hdetect/terminate"/>

        </node>

         <node name="state_estimation_node" pkg="mav_utils" type="state_estimation_node" clear_params="true" output="screen">

			<remap from="odom_px4" to="pilot/local_position/odom" />
			<remap from="global_pose" to="pilot/global_position/global" />
			<remap from="rangefinder" to="rangefinder" />
			<remap from="estimated_odometry" to="odom" />
			<remap from="avoidance_pose" to="mav_avoidance_pose" />
			<param name="rate" value="30.0"/>
			<param name="range_finder_enable" value="false"/>
			<param name="fix_lat" value="$(arg fix_lat)"/>
			<param name="fix_long" value="$(arg fix_long)"/>
        </node>
	    
        <node pkg="mav_utils" type="home_reset.sh" name="home_reset" clear_params="true"/>
        
        <node name="avoidance_node" type="avoidance_node" pkg="avoidance" output="screen">
            <param name="mission_obs" value="true"/>
            <param name="t_to_hit_thresh" value="5"/>
            <param name="min_dist" value="3"/>
            <param name="flight_alt" value="5"/>
            <param name="delta_z" value="0.2"/>
            <param name="mav_id" value="$(arg mav_id)"/>
            <remap from="command/pose" to="pilot/setpoint_position/local"/>
            <remap from="pose_mav1" to="utm_pose"/>
            <remap from="pose_mav2" to="$(arg uav2_utm_pose)"/>
            <remap from="pose_mav3" to="$(arg uav3_utm_pose)"/>
        </node>

<!-- <node name="controller_switch_node" pkg="mav_utils" type="controller_switch_node" clear_params="true" output="screen"/> -->

        <node pkg="mavros" type="mavros_node" name="pilot" required="$(eval not respawn_mavros)" clear_params="true" output="$(arg log_output_mavros)" respawn="$(arg respawn_mavros)">
			<param name="fcu_url" value="$(arg fcu_url_pilot)" />
			<param name="gcs_url" value="$(arg gcs_url_pilot)" />
			<param name="fcu_protocol" value="$(arg fcu_protocol)" />
			<param name="target_system_id" value="$(arg mav_id)" />
			<rosparam command="load" file="$(find mav_utils)/cfg/mavros/px4_pluginlists.yaml" />
			<rosparam command="load" file="$(find mav_utils)/cfg/mavros/px4_config.yaml" />
		</node>

        <node name="router" pkg="imav_planner" type="router" output="screen">

            <!-- parameters -->
            <rosparam file="$(find imav_planner)/cfg/planner.yaml" command="load" ns="" />
            <param name="mav_name" value="$(arg mav_name)"/>

            <!-- subscribers -->
            <remap from="odometry" to="odom"/>
            <remap from="objects" to="object_poses"/>
            <remap from="utm_pose" to="utm_pose"/>
            <remap from="curr_state" to="planner/curr_state"/>
            <remap from="global_pose" to="pilot/global_position/global"/>

            <!-- publishers -->
            <remap from="task" to="drop_info"/>
            <remap from="data" to="data"/>

        </node>

	    <node name="detector" pkg="detector" type="detector_node" output="screen">
            <param name="mav_name" value="$(arg mav_name)"/>
            <rosparam file="$(find detector)/cfg/detector.yaml"/>
            <rosparam file="$(arg camera_info_path)"/>
            <remap from="image" to="$(arg image_topic)"/>
            <remap from="odometry" to="odom"/>
            <remap from="bounding_boxes" to="bounding_boxes"/>
            <remap from="object_poses" to="object_poses"/>
            <remap from="undist_image" to="undist_image"/>
            <remap from="marked_image" to="marked_image"/>
        </node>
	
	    <node pkg="helipad_det" type="hdetect" name="hdetect" output="screen" required="true">
            <rosparam file="$(find helipad_det)/cfg/params.yaml"/>
            <!-- Subscribed Topics -->
            <remap from="odom" to="pilot/local_position/odom"/>
            <remap from="image_raw" to="$(arg image_topic)"/>
            <!-- Published Topics -->
            <remap from="detected_helipad" to="detected_helipad"/>
            <remap from="preprocessed_image" to="preprocessed_image"/>
            <remap from="helipad_position" to="helipad_position"/>
        </node>

    </group>

</launch>